Add `--update` option to BaselineGenerator

dougbu opened the pull request:
- aspnet/AspNetCore-Internal#1649

dougbu committed:
Add `--update` option to BaselineGenerator
- aspnet/AspNetCore-Internal#1649

dougbu commented on eng/tools/BaselineGenerator/BaselineGenerator.csproj:
```diff
@@ -5,10 +5,11 @@
 
   <ItemGroup>
-    <PackageReference Include="NuGet.Packaging" Version="4.8.0" />
+    <PackageReference Include="NuGet.Protocol" Version="4.8.2" />
```
@Eilon please let me know if bumping the NuGet.Packaging version up a couple of patches or expanding the dependencies to include NuGet.Protocol is an issue.

Eilon replied:
Not an issue, thanks for checking.

Eilon replied:
(At least, I don't have an issue from a policy perspective. Not sure if there's any other kind of problem.)

natemcmaster commented on eng/tools/BaselineGenerator/BaselineGenerator.csproj:
```diff
@@ -5,10 +5,11 @@
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <StartArguments>-o "$(MSBuildThisFileDirectory)../../Baseline.Designer.props"</StartArguments>
     <StartWorkingDirectory>$(MSBuildThisFileDirectory)../../</StartWorkingDirectory>
+    <LangVersion>preview</LangVersion>
```
> CSC : error CS1617: Invalid option 'preview' for /langversion. Use '/langversion:?' to list supported values. [D:\a\1\s\eng\tools\BaselineGenerator\BaselineGenerator.csproj]


Which version are you trying to use? I think the latest version supported by the 2.1 SDK is C# 7.3

dougbu replied:
Hmm, I was actually building locally w/ the 2.2.203 SDK. I'll add a `global.json` and see if VS complains about my use of `using var`

dougbu replied:
Either way, the Code Check step is failing w/ `CSC : error CS1617: Invalid option 'preview' for /langversion. Use '/langversion:?' to list supported values. [D:\a\1\s\eng\tools\BaselineGenerator\BaselineGenerator.csproj]`

natemcmaster replied:
using var is a c# 8 feature, which our 2.x branches don't support. You will have to stick to c# 7

dougbu committed:
Stop using C# 8 features and correct execution issues
- find unlisted packages when using `--update`
- correct `FloatRange` minimum version
- remove `$(StartArguments}` from project; conflicts with `--update`
- support v3 source _automatically_ when not using `--update`

nits:
- `--source` -> `--packageSource` due to conflict with a `dotnet` argument
- write out Baseline.xml file in UTF8
- grab a couple of improvements from 'release/2.2'
  - never exit silently
  - remove `first` special case

dougbu commented:
:up::date: to get things compiling and working
- also added automatic detection of a V3 source

dougbu commented on eng/tools/BaselineGenerator/Program.cs:
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
         {
             var source = _source.HasValue()
-                ? _source.Value()
-                : "https://www.nuget.org/api/v2/package";
```
- This value wasn't an actual package source anyhow
- Should we use the V3 feed by default?

natemcmaster replied:
> Should we use the V3 feed by default?

Sure, if it works.

dougbu committed:
Update README.md

dougbu committed:
Add sentence from Extensions version of README.md

natemcmaster commented on eng/tools/BaselineGenerator/BaselineGenerator.csproj:
```diff
@@ -3,12 +3,11 @@
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.1</TargetFramework>
-    <StartArguments>-o "$(MSBuildThisFileDirectory)../../Baseline.Designer.props"</StartArguments>
```
Why is this removed?

dougbu replied:
1. It conflicts with `--update`
2. It's not useful because `$(StartWorkingDirectory)` sets the current directory correctly and the tool defaults to finding `Baseline.Designer.props` in its current directory

natemcmaster commented on eng/tools/BaselineGenerator/Program.cs:
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
         public Program()
         {
-            _source = Option("-s|--source <SOURCE>", "The NuGet v2 source of the package to fetch", CommandOptionType.SingleValue);
+            _source = Option("-s|--packageSource <SOURCE>", "The NuGet source of packages to fetch", CommandOptionType.SingleValue);
```
```suggestion
            _source = Option("-s|--package-source <SOURCE>", "The NuGet source of packages to fetch", CommandOptionType.SingleValue);
```

Nit: follow POSIX conventions for capitalization of command line args

natemcmaster commented on eng/tools/BaselineGenerator/README.md:
```diff
@@ -1,10 +1,20 @@
 
-1. Add to the [Baseline.xml](/eng/Baseline.xml) a list of package ID's and their latest released versions. The source of this information can typically
-  be found in the build.xml file generated during ProdCon builds. See https://github.com/dotnet/versions/blob/master/build-info/dotnet/product/cli/release/2.1.6/build.xml for example.
+Add `--packageSource {source}` to the commands below if the packages of interest are not all hosted on NuGet.org.
```
```suggestion
Add `--package-source {source}` to the commands below if the packages of interest are not all hosted on NuGet.org.
```

natemcmaster reviewed (approved):
A few nit's and one question

dougbu commented on eng/tools/BaselineGenerator/Program.cs:
```diff
@@ -26,34 +32,57 @@ static void Main(string[] args)
-                ? Environment.GetEnvironmentVariable("NUGET_PACKAGES")
-                : Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget", "packages");
+                ? _source.Value().TrimEnd('/')
+                : "https://www.nuget.org/api/v2";
```
```suggestion
                : "https://api.nuget.org/v3/index.json";
```

dougbu committed:
Use V3 source by default

natemcmaster committed:
Apply suggestions from code review

- `--packageSource` -> `--package-source`

Co-Authored-By: dougbu <drb24@cornell.edu>
